File: day-12-graphs-bfs/js/bfs.js
// The graph
const graph = {
  you: ["alice", "bob", "claire"],
  alice: ["peggy"],
  bob: ["anuj", "peggy"],
  claire: ["thom", "jonny"],
  anuj: [],
  peggy: [],
  thom: [],
  jonny: [],
};

// Person we look for
function isMangoSeller(name) {
  return name[name.length - 1] === "m"; // Silly rule!
}

function bfs(graph, start) {
  const queue = [...graph[start]];
  const visited = new Set();

  while (queue.length > 0) {
    const person = queue.shift();
    if (!visited.has(person)) {
      visited.add(person);
      if (isMangoSeller(person)) {
        return person; // Found!
      }
      // Add their friends to the queue
      queue.push(...graph[person]);
    }
  }
  return null; // No seller!
}

console.log(bfs(graph, "you"));

File: day-12-graphs-bfs/python/bfs.py
from collections import deque

# Graph
graph = {
    "you": ["alice", "bob", "claire"],
    "alice": ["peggy"],
    "bob": ["anuj", "peggy"],
    "claire": ["thom", "jonny"],
    "anuj": [],
    "peggy": [],
    "thom": [],
    "jonny": []
}

def is_mango_seller(name):
    return name[-1] == "m" # Silly rule!

def bfs(graph, start):
    queue = deque(graph[start]) # start with the frist-degree friends
    visited = set()

    while queue:
        person = queue.popleft() # dequeue from the left. using pop(0) is O(n), so we avoid that
        if person not in visited:
            visited.add(person)
            if is_mango_seller(person):
                return person # Found!
            queue.extend(graph[person])
        
    return None

print(bfs(graph, "you"))